### DISCUSSION TITLE
Is it possible to only build meshlabserver?

### DISCUSSION THREAD

**Discussion by Sea421** [cnr-isti-vclab/meshlab#871] [Q&A] [answered]
Hi team, 

I would like you use meshlabserver on my server to convert obj file to u3d. 
I just wondering is it possible to only build or install meshlabserver without whole meshlab?

Thanks

**Comment by alemuntoni** [comment, marked as answer]
MeshLab has a plugin architecture, therefore you could build just what you need;

- external/u3d
- common
- meshlabserver
- meshlabplugins/io_u3d

Cmake allows to select which targets to build.

---

However, to convert u3d files (until u3d exporter will not work in PyMeshLab), I suggest to use the following pipeline:

- convert your file to IDTF format using pymeshlab
- use [this converter](https://github.com/ningfei/u3d) to convert from IDTF to u3d.

**Reply by Sea421** [reply]
Thank you very much. u3d conversion works fine on my local ubuntu 18.04. (pymeshlab v0.1.16)
But in docker (ubuntu 18.04), it returns following error:
``` 
Must construct a QGuiApplication first.
Exit code = 80000000
QWidget: Must construct a QApplication before a QWidget
Aborted (core dumped)
```
I have install u3d on my local machine and docker. But it works fine in my local machine, also give` Exit code = 80000000` error on docker.

Do I need to set configuration on server, or is there headless version of pymeshlab on server? 

My code:

``` python
import pymeshlab as ml

ms = ml.MeshSet()
ms.load_new_mesh("my_file.ply")
ms.save_current_mesh("my_file.u3d")
ms.save_current_mesh("my_file.idtf")
```


``` 
IDTFConverter -input my_file.idtf -output my_file.u3d
```

Thanks again

**Reply by alemuntoni** [reply]
I really don't know why the u3d converter does not work on server. 
Be sure that, with the converter, also the other IFX libraries are in the same folder of the converter.
Unfortunately the IDTF converter is a total mess that loads internally the libraries instead of linking them, therefore something may go wrong during this loading... See https://github.com/cnr-isti-vclab/PyMeshLab/discussions/25

**Reply by Sea421** [reply]
I have solved the issue. The issue was IDTFConverter is using the en_US.UTF-8 locale. But, in my docker container, the en_US.UTF-8 locale was missing. So, I installed the locale.
```
apt-get -y install locales
sed -i '/en_US.UTF-8/s/^# //g' /etc/locale.gen && locale-gen
```

For, u3d library, the error occurred because of it can't find .so files. So I have copied built so files to /usr/lib/
```
cp libIDTF.so /usr/local/lib/
cp libIFXCoreso /usr/local/lib/
cp libIFXExporting.so /usr/local/lib/
cp libIFXImporting.so /usr/local/lib/
cp libIFXScheduling.so /usr/local/lib/
```

Thanks for great tool pymeshlab & meshlab. It is really saving my work.

**Reply by ghost** [reply]
@Sea421, thanks for the tip! It was indeed the missing locale setting that was breaking IDTFConverter on the Ubuntu image.
